### PR TITLE
Thumbnail for links

### PR DESCRIPTION
**Pull request by gempir**
# Description
With https://github.com/Chatterino/api/pull/14 I'm adding a thumbnail to the response of the link resolver. 

In this PR I'm adding this functionality to chatterino so you can preview links. This will definetely be off by default in the settings. 

I'm still trying to figure out how to resize the image (set a max size) any pointers would be great! 

Also I wanted to open this Draft to pose the the question if it's okay that this can potentially lead to a IP leak when someone posts a valid img link and you hover over it with the setting enabled. 

If that's not okay we either need to proxy images, which I think is a bit much. Or we only allow thumbnails for trusted urls? Not sure if that list should be in chatterino or the api. 

image

# TODOs

- [x] add max size for thumbnail
- [x] introduce setting for thumbnail (should be off by default)
- [x] think about IP leaking, since all urls with an image would return a thumbnail not hosted on chatterino api
- [x] update thumbnail size on config update

### CODE REVIEW

**Comment by alazymeme** [comment]
I like this :+1:

**Comment by Mm2PL** [comment]
Maybe the IP leak problem could be semi-solved using an option.
By default we could allow only certain services, like `imgur` or `i.nuuls.com`, the user could then decide to load more thumbnails.

**Comment by gempir** [comment]
We will solve this IP leak issue by caching all images besides YT/Twitch Clips in the chatterino-api

**Comment by gempir** [comment]
This is ready for review now. Requires https://github.com/Chatterino/api/pull/14 to show any thumbnails.

API only supports gif, png and jpeg. Gifs are static for now, I might fix them at a later date in another PR.

And it looks something like this 

image

File: src/singletons/Settings.hpp [hunk 1]
```diff
@@ -161,6 +161,7 @@ class Settings : public ABSettings, public ConcurrentSettings
     /// Links
     BoolSetting linksDoubleClickOnly = {"/links/doubleClickToOpen", false};
     BoolSetting linkInfoTooltip = {"/links/linkInfoTooltip", false};
+    FloatSetting thumbnailSize = {"/appearance/thumbnailSize", 0};
```
Comment (pajlada) [inline src/singletons/Settings.hpp:164]: Can we make this an IntSetting instead of a FloatSetting

File: src/widgets/helper/ChannelView.cpp [hunk 2]
```diff
@@ -1288,7 +1289,22 @@ void ChannelView::mouseMoveEvent(QMouseEvent *event)
         }
         else
         {
-            tooltipPreviewImage.setImage(nullptr);
+            auto element = &hoverLayoutElement->getCreator();
+            auto thumbnailSize = getSettings()->thumbnailSize;
+            if (thumbnailSize == 0.f)
```
Comment (pajlada) [inline src/widgets/helper/ChannelView.cpp:1294]: remember to modify this when we change the setting from a float to an int

File: src/widgets/settingspages/GeneralPage.cpp [hunk 3]
```diff
@@ -521,6 +521,31 @@ void GeneralPage::initLayout(SettingsLayout &layout)
...
+        },
+        [](auto args) {
+            if (args.value == "Small")
+                return 100.f;
+            else if (args.value == "Medium")
+                return 200.f;
+            else if (args.value == "Large")
+                return 300.f;
+
+            return fuzzyToFloat(args.value, 0.f);
```
Comment (pajlada) [inline src/widgets/settingspages/GeneralPage.cpp:547]: remember to modify this when we change the setting from a float to an int

**Pull request merged**